src/Shared/Microsoft.AspNetCore.Razor.Utilities.Shared/PooledObjects/PooledHashSet`1.cs

```diff
@@ -15,80 +17,228 @@ namespace Microsoft.AspNetCore.Razor.PooledObjects;
 /// </summary>
 internal ref struct PooledHashSet<T>
 {
-    private readonly ObjectPool<HashSet<T>> _pool;
-#pragma warning disable IDE0052 // Used in NET only code below. Called API doesn't exist on framework.
+    private readonly IEqualityComparer<T> _comparer;
+
+    // Used in NET only code below. Called API doesn't exist on framework.
+#pragma warning disable IDE0052
     private readonly int? _capacity;
 #pragma warning restore IDE0052
+
+    private HashSetPool<T>? _pool;
+    private (bool hasValue, T value) _item;
     private HashSet<T>? _set;
 
     public PooledHashSet()
-        : this(pool: null, capacity: null)
     {
+        _comparer = EqualityComparer<T>.Default;
     }
 
-    public PooledHashSet(ObjectPool<HashSet<T>> pool)
-        : this(pool, capacity: null)
+    public PooledHashSet(int capacity)
     {
+        _comparer = EqualityComparer<T>.Default;
+        _capacity = capacity;
     }
 
-    public PooledHashSet(int capacity)
-        : this(pool: null, capacity)
+    public PooledHashSet(IEqualityComparer<T> comparer)
     {
+        ValidateComparer(comparer);
+        _comparer = comparer;
     }
 
-    public PooledHashSet(ObjectPool<HashSet<T>>? pool, int? capacity)
+    public PooledHashSet(IEqualityComparer<T> comparer, int capacity)
     {
-        _pool = pool ?? HashSetPool<T>.Default;
+        ValidateComparer(comparer);
+        _comparer = comparer;
         _capacity = capacity;
     }
 
+    public PooledHashSet(HashSetPool<T> pool, int capacity)
+    {
+        _comparer = pool.Comparer;
+        _pool = pool;
+        _capacity = capacity;
+    }
+
+    public PooledHashSet(HashSetPool<T> pool)
+    {
+        _comparer = pool.Comparer;
+        _pool = pool;
+    }
+
+    [Conditional("DEBUG")]
+    private static void ValidateComparer(IEqualityComparer<T> comparer)
+    {
+        if (!ReferenceEquals(comparer, EqualityComparer<T>.Default) &&
+            !ReferenceEquals(comparer, StringComparer.Ordinal) &&
+            !ReferenceEquals(comparer, StringComparer.OrdinalIgnoreCase))
+        {
+            ThrowHelper.ThrowArgumentException(nameof(comparer),
+                "Only EqualityComparer<T>.Default, StringComparer.Ordinal, and StringComparer.OrdinalIgnoreCase are supported. Please provide a pool.");
+        }
+    }
+
+    private readonly IEqualityComparer<T> Comparer => _comparer ?? EqualityComparer<T>.Default;
+
+    private readonly bool HasSingleItem => _item.hasValue && _set is null;
+
+    [MemberNotNullWhen(false, nameof(_set))]
+    private readonly bool SetIsNullOrEmpty => _set is null || _set.Count == 0;
+
     public void Dispose()
     {
         ClearAndFree();
     }
 
     public readonly int Count
-        => _set?.Count ?? 0;
+    {
+        get
+        {
+            if (HasSingleItem)
```

Comment:
>if (HasSingleItem) [](http://example.com/codeflow?start=12&length=18)

I'm curious whether most users really want to specially handle small hashsets and even the slight overhead added to do so. Would it make sense to have a PooledSmallHashSet struct instead that had this behavior?

Reply:
Makes sense. Thanks!